[indexstore] Remove conditional strong enums from the API

jkorous-apple opened the pull request:
We got a stage2 build failure after this patch landed:
d6425e2c14d6425e2c143 2020-05-10 Richard Smith:  Properly implement 'enum class' parsing.

Since this should remain a pure C API we are sticking to simple enums.

rdar://problem/63156155

jkorous-apple committed:
[indexstore] Remove conditional strong enums from the API

We got a stage2 build failure after this patch landed:
d6425e2c14d6425e2c143 2020-05-10 Richard Smith:  Properly implement 'enum class' parsing.

Since this should remain a pure C API we are sticking to simple enums.

rdar://problem/63156155

jkorous-apple commented:
@swift-ci please test

benlangmuir commented:
@kastiglione added these annotations

jkorous-apple commented:
@kastiglione I can't add you as a reviewer but please feel free to give feedback anyway

kastiglione commented:
fwiw the indexstore API has optional support for blocks, so it's not pure C even if this is removed.

Do you have a link to the build error? Is there a way to improve the macros so that this can be kept while also fixing the build? This code improves interop/bridging from Swift, it would be nice if there were a way forward that preserved it.

akyrtzi commented:
Is the Swift bridging benefiting from `objc_fixed_enum` while the issue is coming from `cxx_strong_enums`? If yes, perhaps only remove the latter?

jkorous-apple commented:
The error is:

```
While building module 'Clang_Index' imported from /jenkins/workspace/apple-clang-stage2-configure-RA_master/llvm-project/clang/lib/CrossTU/CrossTranslationUnit.cpp:20:
While building module 'Clang_Index_C' imported from /jenkins/workspace/apple-clang-stage2-configure-RA_master/llvm-project/clang/include/clang/Index/IndexDataStoreSymbolUtils.h:13:
In file included from <module-includes>:1:
/jenkins/workspace/apple-clang-stage2-configure-RA_master/llvm-project/clang/include/indexstore/indexstore.h:290:9: error: non-defining declaration of enumeration with a fixed underlying type is only permitted as a standalone declaration; missing list of enumerators? [-Welaborated-enum-base]
typedef INDEXSTORE_OPTIONS(uint64_t, indexstore_symbol_property_t) {
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/jenkins/workspace/apple-clang-stage2-configure-RA_master/llvm-project/clang/include/indexstore/indexstore.h:103:79: note: expanded from macro 'INDEXSTORE_OPTIONS'
# define INDEXSTORE_OPTIONS(_type, _name) enum INDEXSTORE_OPTIONS_ATTRS _name : _type _name; enum INDEXSTORE_OPTIONS_ATTRS _name : _type
                                                                              ^~~~~~~
```

jkorous-apple commented:
Richard's patch seems to be only dealing with C++ enums. I can try to preserve conditional behavior for `objc_fixed_enum`.

kastiglione commented:
from the change (relevant excerpt below), and the above diagnostic, it looks like the declaration shouldn't have the underlying type, only the definition should.

from the commit message:

> Bumped the recently-added `ExtWarn` for `enum E : int x;` to be `DefaultError`;

from the patch:

```diff
+  // An elaborated-type-specifier has a much more constrained grammar:
+  //
+  //   'enum' nested-name-specifier[opt] identifier
+  //
+  // If we parsed any other bits, reject them now.
+  //
+  // MSVC and (for now at least) Objective-C permit a full enum-specifier
+  // or opaque-enum-declaration anywhere.
+  if (IsElaboratedTypeSpecifier && !getLangOpts().MicrosoftExt &&
+      !getLangOpts().ObjC) {
```

kastiglione commented:
The c++ portion of the macro was included to match the equivalent implementation in `CFAvailability.h` and `os.h`.

From the diagnostic, it looks like the fix could be changing the declaration:
`enum INDEXSTORE_OPTIONS_ATTRS _name : _type _name;`

to:
`enum INDEXSTORE_OPTIONS_ATTRS _name _name;`

Removing the c++ bits entirely also works for my case I think.

jkorous-apple commented:
You mean this change, correct?

`# define INDEXSTORE_OPTIONS(_type, _name) enum INDEXSTORE_OPTIONS_ATTRS _name : _type _name; enum INDEXSTORE_OPTIONS_ATTRS _name : _type`

->

`# define INDEXSTORE_OPTIONS(_type, _name) enum INDEXSTORE_OPTIONS_ATTRS _name : _type; enum INDEXSTORE_OPTIONS_ATTRS _name : _type`

That seems to work.

kastiglione commented:
I had a mistake in my comment (which I've corrected).

from the diagnostic:

```
enum INDEXSTORE_OPTIONS_ATTRS _name : _type _name;
                                    ^~~~~~~
```

The error points to the `: _type` part in the declaration.

The change you've shown removes `_name` from the declaration. It appears it's the type not the name that should be removed.

jkorous-apple commented:
Can you please check this is what you meant?

With this macro:
`# define INDEXSTORE_OPTIONS(_type, _name) enum INDEXSTORE_OPTIONS_ATTRS _name _name; enum INDEXSTORE_OPTIONS_ATTRS _name : _type`

I get build errors (stage 1 - without Richard's patch) like this one:
`/Users/jankorous/src/llvm/internal/master/worktree/clang/include/indexstore/indexstore.h:290:38: error: ISO C++ forbids forward references to 'enum' types
typedef INDEXSTORE_OPTIONS(uint64_t, indexstore_symbol_property_t) {
                                     ^
/Users/jankorous/src/llvm/internal/master/worktree/clang/include/indexstore/indexstore.h:290:38: error: enumeration previously declared with nonfixed underlying type
/Users/jankorous/src/llvm/internal/master/worktree/clang/include/indexstore/indexstore.h:290:38: note: previous declaration is here`

The relevant part of preprocessed `indexstore.h` is:
`typedef enum __attribute__((enum_extensibility(open))) __attribute__((flag_enum)) indexstore_symbol_property_t indexstore_symbol_property_t; enum __attribute__((enum_extensibility(open))) __attribute__((flag_enum)) indexstore_symbol_property_t : uint64_t {
  INDEXSTORE_SYMBOL_PROPERTY_GENERIC = 1 << 0,
// ...`

kastiglione commented:
ok, that doesn't work.

Maybe removing the c++ part `__has_extension(cxx_strong_enums)` is the way to go.

jkorous-apple commented:
Ok! Thanks

jkorous-apple committed:
Revert "[indexstore] Remove conditional strong enums from the API"

This reverts commit e0993b3e24d41663f4b62cce9050af60055c1872.

jkorous-apple committed:
[indexstore] Preserve enums in the API for Obj-C fixed enums

It turned out this is important for Swift - Obj-C interop.

Relevant discussion here:
https://github.com/apple/llvm-project/pull/1229

jkorous-apple commented:
@swift-ci please test

jkorous-apple commented:
@swift-ci please test

kastiglione commented:
thank you @jkorous-apple
